srml/staking/src/lib.rs

```diff
@@ -640,57 +696,141 @@ impl<T: Trait> Module<T> {
 	fn select_validators() -> BalanceOf<T> {
 		// Map of (would-be) validator account to amount of stake backing it.
 		
-		// First, we pull all validators, together with their stash balance into a Vec (cpu=O(V), mem=O(V))
-		let mut candidates = <Validators<T>>::enumerate()
-			.map(|(who, _)| {
-				let stash_balance = Self::stash_balance(&who);
-				(who, Exposure { total: stash_balance, own: stash_balance, others: vec![] })
-			})
-			.collect::<Vec<(T::AccountId, Exposure<T::AccountId, BalanceOf<T>>)>>();
-		// Second, we sort by accountid (cpu=O(V.log(V)))
-		candidates.sort_unstable_by_key(|i| i.0.clone());
-		// Third, iterate through nominators and add their balance to the first validator in their approval
-		// list. cpu=O(N.log(V))
-		for (who, nominees) in <Nominators<T>>::enumerate() {
-			// For this trivial nominator mapping, we just assume that nominators always
-			// have themselves assigned to the first validator in their list.
-			if nominees.is_empty() {
-				// Not possible, but we protect against it anyway.
-				continue;
+		let rounds = <ValidatorCount<T>>::get() as usize;
```

Comment:
Phragmén should be pulled out into its own module. Use closures in order to avoid relying on a particular set of storage items.